src/adapters/shell.coffee

```diff
@@ -26,41 +25,71 @@ class Shell extends Adapter
     @send envelope, strings...
 
   run: ->
-    stdin = process.openStdin()
-    stdout = process.stdout
-
-    @repl = null
-    Readline.createInterface
-      path: ".hubot_history",
-      input: stdin,
-      output: stdout,
-      maxLength: historySize, # number of entries
-      next: (rl) =>
-        @repl = rl
-        @repl.on 'close', =>
-          stdin.destroy()
-          @robot.shutdown()
-          process.exit 0
-
-        @repl.on 'line', (buffer) =>
-
-          switch buffer.toLowerCase()
-            when "exit"
-              @repl.close()
-            when "history"
-              stdout.write "#{line}\n" for line in @repl.history
-            else
-              user_id = parseInt(process.env.HUBOT_SHELL_USER_ID or '1')
-              user_name = process.env.HUBOT_SHELL_USER_NAME or 'Shell'
-              user = @robot.brain.userForId user_id, name: user_name, room: 'Shell'
-              @receive new TextMessage user, buffer, 'messageId'
-          @repl.prompt(true)
-
-        @emit 'connected'
-
-        @repl.setPrompt "#{@robot.name}> "
-        @repl.prompt(true)
+    @buildCli()
+
+    @loadHistory (history) =>
+      @cli.history(history)
+      @cli.interact("#{@robot.name}> ")
+      @emit 'connected'
+
+  shutdown: () ->
+    @robot.shutdown()
+    process.exit 0
+
+  buildCli: () ->
+    @cli = cline()
+
+    @cli.command '*', (input) =>
+      userId = parseInt(process.env.HUBOT_SHELL_USER_ID or '1')
+      userName = process.env.HUBOT_SHELL_USER_NAME or 'Shell'
+      user = @robot.brain.userForId userId, name: userName, room: 'Shell'
+      @receive new TextMessage user, input, 'messageId'
+
+    @cli.command 'history', () =>
+      console.log item for item in @cli.history()
+
+    @cli.on 'history', (item) =>
+      if item.length > 0 and item isnt 'exit' and item isnt 'history'
+        fs.appendFile '.hubot_history', "#{item}\n", (err) =>
+          @robot.emit 'error', err if err
+
+    @cli.on 'close', () =>
+      history = @cli.history()
+      if history.length > historySize
+        startIndex = history.length - historySize
+        history = history.reverse().splice(startIndex, historySize)
+
+        outstream = fs.createWriteStream('.hubot_history')
+        # >= node 0.10
+        outstream.on 'finish', () =>
+          @shutdown()
+
+        for item in history
+          outstream.write "#{item}\n"
```

Comment:
> I was originally thinking to not join it to avoid keeping it in memory.

That’s legit :+1:

> (ie there's not a trailing newline).

`Array#join` doesn’t add a trailing newline, the separator is just added inbetween the items:

```
coffee> [ 'a', 'b' ].join("\n")
'a\nb'
```